libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs

```diff
@@ -433,6 +449,62 @@ public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext
             return responses;
         }
 
+        /// <summary>
+        /// Forward an activity to a skill(bot).
+        /// </summary>
+        /// <remarks>NOTE: Forwarding an activity to a skill will flush UserState and ConversationState changes so that skill has accurate state.</remarks>
+        /// <param name="turnContext">turnContext.</param>
+        /// <param name="skillId">skillId of the skill to forward the activity to.</param>
+        /// <param name="activity">activity to forward.</param>
+        /// <param name="cancellationToken">cancellation Token.</param>
+        /// <returns>Async task with optional invokeResponse.</returns>
+        public override async Task<InvokeResponse> ForwardActivityAsync(ITurnContext turnContext, string skillId, Activity activity, CancellationToken cancellationToken)
+        {
+            // route the activity to the skill
+            var skill = Skills.FirstOrDefault(s => s.Name == skillId);
+            if (skill != null)
+            {
+                activity.Recipient.Properties["skillId"] = skill.Name;
+
+                // Encode original bot service URL and ConversationId in the new conversation ID so we can unpack it later.
+                
+                // TODO use SkillConversation class here instead of hard coded encoding...
+                // var skillConversation = new SkillConversation() { ServiceUrl = activity.ServiceUrl, ConverationId = activity.Conversation.Id };
+                // activity.Conversation.Id = skillConversation.GetSkillConverationId()
+                activity.Conversation.Id = Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(new string[] { activity.Conversation.Id, activity.ServiceUrl })));
+                activity.ServiceUrl = SkillsCallbackUri.ToString();
```

Comment:
If schema could be modified , I would suggest to change ServiceUrl to a stack.

Then when bot forwards message to another bot, it pushes its callback uri on ServiceUrl, and when it replies, it replies to the top of the stack and pop.

In this way, we could support a chain of bots instead of only two.

Reply:
Why I become a reviewer.. Strange.

I am also curious about remote auth: how do we know if it is a IRemoteUserTokenProvider or IUserTokenProvider?